core/src/main/scala/org/apache/spark/eventhubs/EventHubsConf.scala

```diff
@@ -561,6 +568,98 @@ final class EventHubsConf private (private val connectionStr: String)
             s"${PartitionPreferredLocationStrategy.values.mkString(",")}"))
   }
 
+  /**
+   * Set whether to use AAD auth to connect eventhubs.
+   * More info about this: https://docs.microsoft.com/en-us/azure/event-hubs/authorize-access-azure-active-directory
+   * There're 2 options for AadAuth: 1) Certificate; 2) Secret, you'll also need to set
```

Comment:
Instead of asking users to provide certificate or secret as an input, what about allowing users to provide the implementation of AzureActiveDirectoryTokenProvider.AuthenticationCallback API so that the library can remain flexible in terms of supporting various ways of obtaining an access token?

Reply:
good idea!  we can do that, but on the other side, according to azure AD's doc, for app/sp, it only supports either secret or cert to authenticate: https://docs.microsoft.com/en-us/azure/active-directory/develop/quickstart-register-app.
All ehs configuration parameters are String, if we want to keep consistent with others and use an interface implementation as a parameter, we may have to use reflection to create/call the method in runtime. I can add one more option in "AadAuth" to do that if you prefer, but I highly suspect whether it'll be really used in the feature. :)

Reply:
Hi duhuan - thank you for your reply. There is a similar implementation done for metric plug-in. So, one option could be to expose a configuration which a user can set for their implementation of an interface that contains the callback API (https://docs.microsoft.com/en-us/java/api/com.microsoft.azure.servicebus.security.azureactivedirectorytokenprovider.authenticationcallback?view=azure-java-stable).

Reply:
yeah, just added the callback support. please see the latest change :)